packages/dcos-integration-test/extra/test_mesos.py

```diff
@@ -325,3 +328,132 @@ def test_fault_domain(dcos_api_session):
 
         # agent_zone might be different on agents, so we just make sure it's a sane value
         assert agent_zone, 'agent_zone cannot be empty'
+
+
+@pytest.fixture
+def reserved_disk(dcos_api_session):
+    """
+    Set up an agent with one disk in a role.
+
+    Reserve a chunk of `disk` resources on an agent for a role, and the
+    remaining resources to another role. With that a framework in the first
+    role will only be offered `disk` resources.
+    """
+    # Setup.
+
+    def principal():
+        is_enterprise = os.getenv('DCOS_ENTERPRISE', 'false').lower() == 'true'
+
+        if is_enterprise:
+            return dcos_api_session.auth_user.uid
+        else:
+            return 'reserved_disk_fixture_principal'
+
+    dcos_api_session.principal = principal()
+
+    # Get the ID of a private agent. We some assume that resources on that
+    # agent are unreserved.
+    r = dcos_api_session.get('/mesos/slaves')
+    assert r.status_code == 200, r.text
+    response = json.loads(r.text)
+    slaves = [
+        slave['id'] for slave in response['slaves']
+        if 'public_ip' not in slave['attributes']]
+    assert slaves, 'Could not find any private agents'
+    slave_id = slaves[0]
+
+    # Create a unique role to reserve the disk to. The test framework should
+    # register in this role.
+    dcos_api_session.role = 'disk-' + uuid.uuid4().hex
+
+    resources1 = {
+        'agent_id': {'value': slave_id},
+        'resources': [
+            {
+                'type': 'SCALAR',
+                'name': 'disk',
+                'reservations': [
+                    {
+                        'type': 'DYNAMIC',
+                        'role': dcos_api_session.role,
+                        'principal': dcos_api_session.principal,
+                    }
+                ],
+                'scalar': {'value': 32}
+            }
+        ]
+    }
+
+    request = {'type': 'RESERVE_RESOURCES', 'reserve_resources': resources1}
+    r = dcos_api_session.post('/mesos/api/v1', json=request)
+    assert r.status_code == 202, r.text
```

Comment:
Not sure about these assertions in the fixture. If one of them fails, we won't clean up the fixture and not unreserve.

Reply:
I discussed possible approaches with @gpaul and have updated the code so we clean up as much as possible. Now on to another set of CI run and round of reviews.